sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+        ///   running.
+        /// </remarks>
+        ///
+        public virtual async Task StartAsync()
+        {
+            if (ActiveLoadBalancingTask == null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (ActiveLoadBalancingTask == null)
+                    {
+                        // We expect the token source to be null, but we are playing safe.
+
+                        RunningTaskTokenSource?.Cancel();
+                        RunningTaskTokenSource = new CancellationTokenSource();
+
+                        InstanceOwnership = new Dictionary<string, PartitionOwnership>();
+                        ActivePartitionProcessors = new ConcurrentDictionary<string, Task>();
+                        ActivePartitionProcessorTokenSources = new ConcurrentDictionary<string, CancellationTokenSource>();
+
+                        // Start the main running task.  It is responsible for managing the active partition processing tasks and
+                        // for partition load balancing among multiple event processor instances.
+
+                        ActiveLoadBalancingTask = RunAsync(RunningTaskTokenSource.Token);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Stops the event processor.  In case it isn't running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        /// <remarks>
+        ///   If overridden, the base class implementation must be explicitly called in order to make the event processor stop
+        ///   running.
+        /// </remarks>
+        ///
+        public virtual async Task StopAsync()
+        {
+            if (ActiveLoadBalancingTask != null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (ActiveLoadBalancingTask != null)
+                    {
+                        // Cancel the current running task.
+
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        // Now that a cancellation request has been issued, wait for the running task to finish.  In case something
+                        // unexpected happened and it stopped working midway, this is the moment we expect to catch an exception.
+
+                        try
+                        {
+                            await ActiveLoadBalancingTask.ConfigureAwait(false);
+                        }
+                        catch (TaskCanceledException)
```

Comment:
May also be an OperationCanceled, those are often not fully consistent.  How sure are you that it's always TaskCanceled?

Reply:
I'm not sure. I was tired.